File: submodules/android-store/submodules/soomla-android-core/src/com/soomla/data/KeyValueStorage.java
/*
 * Copyright (C) 2012-2014 Soomla Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.soomla.data;

import android.text.TextUtils;

import com.soomla.SoomlaApp;
import com.soomla.SoomlaConfig;
import com.soomla.SoomlaUtils;
import com.soomla.util.AESObfuscator;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

/**
 * This class provides basic storage operations for a simple key-value store.
 */
public class KeyValueStorage {

    /**
     * Retrieves the value for the given key.
     *
     * @param key is the key in the key-val pair
     * @return the value for the given key
     */
    public static String getValue(String key) {
        SoomlaUtils.LogDebug(TAG, "trying to fetch a value for key: " + key);

        key = getAESObfuscator().obfuscateString(key);

        String val = getDatabase().getKeyVal(key);

        if (val != null && !TextUtils.isEmpty(val)) {
            try {
                val = getAESObfuscator().unobfuscateToString(val);
            } catch (AESObfuscator.ValidationException e) {
                SoomlaUtils.LogError(TAG, e.getMessage());
                val = "";
            }

            SoomlaUtils.LogDebug(TAG, "the fetched value is " + val);
        }
        return val;
    }

    /**
     * Sets key-val pair in the database according to given key and val.
     *
     * @param key key to set in pair
     * @param val value to set in pair
     */
    public static void setNonEncryptedKeyValue(String key, String val) {
        SoomlaUtils.LogDebug(TAG, "setting " + val + " for key: " + key);

        val = getAESObfuscator().obfuscateString(val);

        getDatabase().setKeyVal(key, val);
    }

    /**
     * Deletes key-val pair that has the given key.
     *
     * @param key the key to indicate which pair to delete
     */
    public static void deleteNonEncryptedKeyValue(String key) {
        SoomlaUtils.LogDebug(TAG, "deleting " + key);

        getDatabase().deleteKeyVal(key);
    }

    /**
     * Retrieves the value of the key-val pair with the given key.
     *
     * @param key key according to which val will be retrieved
     * @return value of key-val pair
     */
    public static String getNonEncryptedKeyValue(String key) {
        SoomlaUtils.LogDebug(TAG, "trying to fetch a value for key: " + key);

        String val = getDatabase().getKeyVal(key);

        if (val != null && !TextUtils.isEmpty(val)) {
            try {
                val = getAESObfuscator().unobfuscateToString(val);
            } catch (AESObfuscator.ValidationException e) {
                SoomlaUtils.LogError(TAG, e.getMessage());
                val = "";
            }

            SoomlaUtils.LogDebug(TAG, "the fetched value is " + val);
        }
        return val;
    }

    /**
     * Retrieves key-val pairs according to given query.
     *
     * @param query query that determines what key-val pairs will be returned
     * @return hashmap of key-val pairs
     */
    public static HashMap<String, String> getNonEncryptedQueryValues(String query) {
        SoomlaUtils.LogDebug(TAG, "trying to fetch values for query: " + query);

        HashMap<String, String> vals = getDatabase().getQueryVals(query);
        HashMap<String, String> results = new HashMap<String, String>();
        for(String key : vals.keySet()) {
            String val = vals.get(key);
            if (val != null && !TextUtils.isEmpty(val)) {
                try {
                    val = getAESObfuscator().unobfuscateToString(val);
                    results.put(key, val);
                } catch (AESObfuscator.ValidationException e) {
                    SoomlaUtils.LogError(TAG, e.getMessage());
                }
            }
        }

        SoomlaUtils.LogDebug(TAG, "fetched " + results.size() + " results");

        return results;
    }

    /**
     * Retrieves one key-val according to given query.
     *
     * @param query query that determines what key-val will be returned
     * @return string of key-val returned
     */
    public static String getOneForNonEncryptedQuery(String query) {
        SoomlaUtils.LogDebug(TAG, "trying to fetch one for query: " + query);

        String val = getDatabase().getQueryOne(query);
        if (val != null && !TextUtils.isEmpty(val)) {
            try {
                val = getAESObfuscator().unobfuscateToString(val);
                return val;
            } catch (AESObfuscator.ValidationException e) {
                SoomlaUtils.LogError(TAG, e.getMessage());
            }
        }

        return null;
    }

    /**
     * Retrieves the number key-vals according to given query.
     *
     * @param query query that determines what number of key-vals
     * @return number of key-vals according the the given query
     */
    public static int getCountForNonEncryptedQuery(String query) {
        SoomlaUtils.LogDebug(TAG, "trying to fetch count for query: " + query);

        return getDatabase().getQueryCount(query);
    }

    /**
     * Gets all keys in the storage with no encryption
     *
     * @return a List of unencrypted keys
     */
    public static List<String> getEncryptedKeys() {
        SoomlaUtils.LogDebug(TAG, "trying to fetch all keys");

        List<String> encryptedKeys = getDatabase().getAllKeys();
        List<String> resultKeys = new ArrayList<String>();

        for (String encryptedKey : encryptedKeys) {
            try {
                String unencryptedKey = getAESObfuscator().unobfuscateToString(encryptedKey);
                resultKeys.add(unencryptedKey);
            } catch (AESObfuscator.ValidationException e) {
                SoomlaUtils.LogDebug(TAG, e.getMessage());
            } catch (RuntimeException e) {
                SoomlaUtils.LogError(TAG, e.getMessage());
            }
        }

        return resultKeys;
    }

    /**
     * Sets the given value to the given key.
     *
     * @param key is the key in the key-val pair.
     * @param val is the val in the key-val pair.
     */
    public static void setValue(String key, String val) {
        SoomlaUtils.LogDebug(TAG, "setting " + val + " for key: " + key);

        key = getAESObfuscator().obfuscateString(key);
        val = getAESObfuscator().obfuscateString(val);

        getDatabase().setKeyVal(key, val);
    }

    /**
     * Deletes a key-val pair with the given key.
     *
     * @param key is the key in the key-val pair.
     */
    public static void deleteKeyValue(String key) {
        SoomlaUtils.LogDebug(TAG, "deleting " + key);

        key = getAESObfuscator().obfuscateString(key);

        getDatabase().deleteKeyVal(key);
    }

    /**
     * Purges the entire storage
     *
     * NOTE: Use this method with care, it will erase all user data in storage
     * This method is mainly used for testing.
     */
    public static void purge() {
        SoomlaUtils.LogDebug(TAG, "purging database");

        getDatabase().purgeDatabaseEntries(SoomlaApp.getAppContext());
    }

    /**
     * Retrieves the key-val database.
     *
     * @return key-val database
     */
    private static synchronized KeyValDatabase getDatabase(){

        if (mKvDatabase == null) {
            mKvDatabase = new KeyValDatabase(SoomlaApp.getAppContext());
        }

        return mKvDatabase;
    }

    /**
     * Retrieves AESObfuscator
     *
     * @return AESObfuscator
     */
    private static AESObfuscator getAESObfuscator(){
        if (mObfuscator == null) {
            mObfuscator = new AESObfuscator(SoomlaConfig.obfuscationSalt,
                    SoomlaApp.getAppContext().getPackageName(), SoomlaUtils.deviceId());
        }

        return mObfuscator;
    }


    /** Private Members **/

    private static final String TAG = "SOOMLA KeyValueStorage"; //used for Log Messages

    private static AESObfuscator mObfuscator;

    private static KeyValDatabase mKvDatabase;
}

File: submodules/soomla-unity3d-core/submodules/soomla-ios-core/SoomlaiOSCore/data/KeyValueStorage.m
/*
 Copyright (C) 2012-2014 Soomla Inc.
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
 http://www.apache.org/licenses/LICENSE-2.0
 
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

#import "KeyValueStorage.h"
#import "KeyValDatabase.h"
#import "SoomlaEncryptor.h"
#import "KeyValDatabase.h"
#import "SoomlaConfig.h"
#import "SoomlaUtils.h"

@implementation KeyValueStorage

+ (KeyValDatabase*)kvDatabase {
    static KeyValDatabase* dbInstance;
    if (!dbInstance) {
        dbInstance = [[KeyValDatabase alloc] init];
    }
    return dbInstance;
}

- (id)init{
    self = [super init];
    if (self){
        
    }
    
    return self;
}

+ (NSString*)getValueForKey:(NSString*)key {
    key = [SoomlaEncryptor encryptString:key];
    NSString* val = [[self kvDatabase] getValForKey:key];
    if (val && [val length]>0){
        return [SoomlaEncryptor decryptToString:val];
    }
    
    return NULL;
}

+ (void)setValue:(NSString*)val forKey:(NSString*)key {
    key = [SoomlaEncryptor encryptString:key];
    [[self kvDatabase] setVal:[SoomlaEncryptor encryptString:val] forKey:key];
}

+ (void)deleteValueForKey:(NSString*)key {
    key = [SoomlaEncryptor encryptString:key];
    [[self kvDatabase] deleteKeyValWithKey:key];
}

+ (NSDictionary*)getKeysValuesForNonEncryptedQuery:(NSString*)query {
    NSDictionary* dbResults = [[self kvDatabase] getKeysValsForQuery:query];
    NSMutableDictionary* results = [NSMutableDictionary dictionary];
    NSArray* keys = [dbResults allKeys];
    for (NSString* key in keys) {
        NSString* val = dbResults[key];
        if (val && [val length]>0){
            NSString* valDec = [SoomlaEncryptor decryptToString:val];
            if (valDec && [valDec length]>0){
                [results setObject:valDec forKey:key];
            }
        }
    }
    
    return results;
}

+ (NSArray*)getValuesForNonEncryptedQuery:(NSString*)query {
    NSArray* vals = [[self kvDatabase] getValsForQuery:query];
    NSMutableArray* results = [NSMutableArray array];
    for (NSString* val in vals) {
        if (val && [val length]>0){
            NSString* valDec = [SoomlaEncryptor decryptToString:val];
            if (valDec && [valDec length]>0){
                [results addObject:valDec];
            }
        }
    }
    
    return results;
}

+ (NSString*)getOneForNonEncryptedQuery:(NSString*)query {
    NSString* val = [[self kvDatabase] getOneForQuery:query];
    if (val && [val length]>0){
        NSString* valDec = [SoomlaEncryptor decryptToString:val];
        if (valDec && [valDec length]>0){
            return valDec;
        }
    }
    
    return NULL;
}

+ (int)getCountForNonEncryptedQuery:(NSString*)query {
    return [[self kvDatabase] getCountForQuery:query];
}


+ (NSString*)getValueForNonEncryptedKey:(NSString*)key {
    NSString* val = [[self kvDatabase] getValForKey:key];
    if (val && [val length]>0){
        return [SoomlaEncryptor decryptToString:val];
    }
    
    return NULL;
}

+ (NSArray *)getEncryptedKeys {
    NSArray *encryptedKeys = [[self kvDatabase] getAllKeys];
    NSMutableArray *resultKeys = [NSMutableArray array];
    
    for (NSString *encryptedKey in encryptedKeys) {
        @try {
            NSString *unencryptedKey = [SoomlaEncryptor decryptToString:encryptedKey];
            if (unencryptedKey) {
                [resultKeys addObject:unencryptedKey];
            }
        }
        @catch (NSException *exception) {
            LogDebug(TAG, ([NSString stringWithFormat:@"Exception while decrypting all keys: %@", exception.description]));
        }
    }
    
    return resultKeys;
}

+ (void)setValue:(NSString*)val forNonEncryptedKey:(NSString*)key {
    [[self kvDatabase] setVal:[SoomlaEncryptor encryptString:val] forKey:key];
}

+ (void)deleteValueForNonEncryptedKey:(NSString*)key {
    [[self kvDatabase] deleteKeyValWithKey:key];
}

+ (void)purge {
    [[self kvDatabase] purgeDatabase];
}

static NSString* TAG = @"SOOMLA KeyValueStorage";

@end
